python/activator/activator.py

```diff
@@ -230,10 +230,19 @@ def next_visit_handler() -> Tuple[str, int]:
                     _log.error(f"Failed to match object id '{oid}'")
             subscriber.acknowledge(subscription=subscription.name, ack_ids=ack_list)
 
-        # Got all the snaps; run the pipeline
-        _log.info(f"Running pipeline on group: {expected_visit.group} detector: {expected_visit.detector}")
-        mwi.run_pipeline(expected_visit, expid_set)
-        return "Pipeline executed", 200
+        if expid_set:
+            # Got at least some snaps; run the pipeline.
+            # If this is only a partial set, the processed results may still be
+            # useful for quality purposes.
+            if len(expid_set) < expected_visit.snaps:
+                _log.warning(f"Processing {len(expid_set)} snaps, expected {expected_visit.snaps}.")
+            _log.info(f"Running pipeline on group: {expected_visit.group} "
+                      f"detector: {expected_visit.detector}")
```

Comment:
The logger that we use in pipetask has an "MDC" that lets you assign a dict to the logger such that every log message reports that dict without having to explicitly add it each time. It's why every message includes the dataId. Is that something that would be useful here?

Reply:
I can't find any mention of MDC in the Python `logging` package (am I looking [in the wrong place](https://docs.python.org/3/library/logging.html)?), but in other loggers I've worked with, MDC is attached to a thread or process. That's a problem here because none of the components are actually associated with a particular visit or even detector. Even each {{MiddlewareInterface}} object is subject to arbitrary reuse (perhaps a design flaw we should fix?), which is why the main content of this PR is basically a bunch of safety-checking code.

In addition, my experience with this ticket suggests that the processing flow isn't a simple thread- or process-isolated architecture -- each worker reports exactly the same PID, and the need for the refresh added in 47b6126b13b01419421bfd9fc9dc04027ab14913 suggests that there is more than one Butler acting on each worker repository (when there shouldn't be). I'd be _very_ worried about inconsistent state if we tried to implement MDC without understanding what's going on there.

Reply:
MDC is not a python logging feature. It's something we added to match lsst.log. We added it in `lsst.daf.butler.core.logging.ButlerMDC`. If you are running multi-threaded then the MDC is not going to work though. pipetask always uses one subprocess per job. We didn't put the effort in to make it multi-threaded.

Reply:
@timj : this is not connected to pipetask at all.

Reply:
I wasn't imagining it would be. I was making a general comment about there existing a way to attach metadata to a logger so that every log message gets the content without the person having to explicitly include it in every log message. I was mentioning pipetask as an example of where we use the MDC -- it's where all those dataId messages come from in long-log mode.  If you are using threads then it's not going to help without patching the MDC class to understand threads.

Reply:
Actually, maybe we don't need the homebrew system. We might be able to do exactly what we need if we use a custom handler and `Handler.setFormatter` (with suitable resource guards). That might let us clean up some of the duplicate formatting code while we're at it.

Reply:
MDC-ing the loggers is deferred to [DM-35828](https://jira.lsstcorp.org/browse/DM-35828).